mentat/edit_history.py

```diff
@@ -1,71 +1,88 @@
 
+from mentat.errors import MentatError
+from mentat.parsers.file_edit import FileEdit, Replacement
+from mentat.session_context import SESSION_CONTEXT
+
 
 # All paths should be abs paths
 @attr.define()
 class RenameAction:
     old_file_name: Path = attr.field()
     cur_file_name: Path = attr.field()
 
-    def undo(self) -> str | None:
+    def undo(self) -> FileEdit:
         if self.old_file_name.exists():
-            return colored(
+            raise MentatError(
                 f"File {self.old_file_name} already exists; unable to undo rename from"
-                f" {self.cur_file_name}",
-                color="light_red",
+                f" {self.cur_file_name}"
             )
         else:
             os.rename(self.cur_file_name, self.old_file_name)
+            return FileEdit(
+                file_path=self.old_file_name, rename_file_path=self.cur_file_name
+            )
 
 
 @attr.define()
 class CreationAction:
     cur_file_name: Path = attr.field()
 
-    def undo(self) -> str | None:
+    def undo(self) -> FileEdit:
         if not self.cur_file_name.exists():
-            return colored(
-                f"File {self.cur_file_name} does not exist; unable to delete",
-                color="light_red",
+            raise MentatError(
+                f"File {self.cur_file_name} does not exist; unable to delete"
             )
         else:
             self.cur_file_name.unlink()
+            return FileEdit(file_path=self.cur_file_name, is_creation=True)
 
 
 @attr.define()
 class DeletionAction:
     old_file_name: Path = attr.field()
     old_file_lines: list[str] = attr.field()
 
-    def undo(self) -> str | None:
+    def undo(self) -> FileEdit:
         if self.old_file_name.exists():
-            return colored(
-                f"File {self.old_file_name} already exists; unable to re-create",
-                color="light_red",
+            raise MentatError(
+                f"File {self.old_file_name} already exists; unable to re-create"
             )
         else:
             with open(self.old_file_name, "w") as f:
                 f.write("\n".join(self.old_file_lines))
+            return FileEdit(file_path=self.old_file_name, is_deletion=True)
 
 
 @attr.define()
 class EditAction:
     cur_file_name: Path = attr.field()
     old_file_lines: list[str] = attr.field()
 
-    def undo(self) -> str | None:
+    def undo(self) -> FileEdit:
         if not self.cur_file_name.exists():
-            return colored(
-                f"File {self.cur_file_name} does not exist; unable to undo edit",
-                color="light_red",
+            raise MentatError(
+                f"File {self.cur_file_name} does not exist; unable to undo edit"
             )
         else:
+            new_file_lines = self.cur_file_name.read_text().split("\n")
             with open(self.cur_file_name, "w") as f:
                 f.write("\n".join(self.old_file_lines))
+            return FileEdit(
```

Comment:
It is a bit weird that we turn a fileedit into a historyaction and then back again; maybe there's a better way to do it, but I think this is fine as long as undo and redo are just for a couple of commands. Just commenting so we can think about it in case we want to improve it later.

Reply:
Agreed. I we get rid of the line numbers when switching from fileedit to historyaction, which is counter-productive. I think I'll need to address this for the `sample-command` PR, so I'll make a note to take care of it there.